IpmiFeaturePkg/Library/IpmiBaseLibMm/IpmiBaseLibMm.c

```diff
@@ -0,0 +1,98 @@
+/** @file
+  A Library to support all BMC access via IPMI command during MM Phase.
+
+  @copyright
+  Copyright 1999 - 2021 Intel Corporation. <BR>
+  Copyright (c) Microsoft Corporation
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+**/
+
+#include <PiDxe.h>
+#include <Protocol/IpmiTransportProtocol.h>
+#include <Library/IpmiBaseLib.h>
+#include <Library/MmServicesTableLib.h>
+#include <Library/DebugLib.h>
+
+STATIC IPMI_TRANSPORT  *mIpmiTransport = NULL;
+
+/**
+  Sends a IPMI command to the BMC and returns the response.
+
+  @param[in]      NetFunction       Net function of the command.
+  @param[in]      Command           IPMI command number.
+  @param[in]      CommandData       Command data buffer.
+  @param[in]      CommandDataSize   Size of command data.
+  @param[out]     ResponseData      Response Data buffer.
+  @param[in,out]  ResponseDataSize  Response data buffer size on input, size of
+                                    read data or required size on return.
+
+  @retval   EFI_SUCCESS             Successfully send IPMI command.
+  @retval   EFI_NOT_AVAILABLE_YET   Ipmi interface is not installed yet.
+**/
+EFI_STATUS
+EFIAPI
+IpmiSubmitCommand (
+  IN UINT8       NetFunction,
+  IN UINT8       Command,
+  IN UINT8       *CommandData,
+  IN UINT32      CommandDataSize,
+  OUT UINT8      *ResponseData,
+  IN OUT UINT32  *ResponseDataSize
+  )
+{
+  EFI_STATUS  Status;
+
+  if (mIpmiTransport == NULL) {
+    Status = gMmst->MmLocateProtocol (&gSmmIpmiTransportProtocolGuid, NULL, (VOID **)&mIpmiTransport);

```

Comment:
nit: If this is a soft requirement, it sounds like it might run very early. Would it be better to check if gMmst is not NULL?

Reply:
@kuqin12, the `gMmst` is the `EFI_MM_SYSTEM_TABLE`, which is installed by `StandaloneMmCore` driver before dispatching the `MM_STANDALONE` drivers, and the `StandaloneMmServicesTableLib` constructor has initialized the gMmst, so the `gMmst` must be available to use in `MM_STANDALONE` driver that link to `StandaloneMmServicesTableLib`.

Reply:
I used to think this is linked to any driver or even core in stmm, which could be a problem if the routine will invoke this before memory allocation is ready on ARM64 systems. Your other comment addressed this one actually :)